### PR TITLE
fix(user): remove banned_at when user is unbanned

### PR DESCRIPTION
**Pull request by wescopeland**
This PR resolves an issue where unbanned users do not have their `banned_at` timestamp on their `UserAccounts` record wiped. This leads to some features remaining disabled for those users across the site.

As part of this work, a few pieces of `SetAccountPermissionsJSON()` have been migrated to Eloquent ORM.

### CODE REVIEW

File: app/Helpers/database/user-permission.php [hunk 1]
```diff
@@ -27,13 +28,13 @@ function SetAccountPermissionsJSON(
     $retVal = [];
     sanitize_sql_inputs($actingUser, $targetUser);
 
-    $targetUserData = [];
-    if (!getAccountDetails($targetUser, $targetUserData)) {
+    $targetUserModel = User::where('User', $targetUser)->first();
```
Comment (luchaos) [inline app/Helpers/database/user-permission.php:31]: The `Model` should ve generally omitted imo as it's implicit. Only in cases where it's ambiguous it makes sense to be added - I know we have a lot occurrences of it being in the variable names which could be removed.
Usually it's done where `user` should be called `username`.

**Review by luchaos — APPROVED** [review]

File: app/Helpers/database/user-permission.php (hunk 1 above)
Comment (wescopeland) [reply app/Helpers/database/user-permission.php]: You're right.

In https://github.com/RetroAchievements/RAWeb/pull/2289/commits/463b9be1b9891f78b8bc08235ba8bd0410efe399, I've made these changes:

* The `$actingUser` string arg has been renamed to `$actingUsername`.
* The `$targetUser` string arg has been renamed to `$targetUsername`.
* The `$targetUserModel` variable has been renamed to `$targetUser`.

File: app/Helpers/database/user-permission.php [hunk 2]
```diff
@@ -19,48 +20,48 @@ function getUserPermissions(?string $user): int
...
-    string $actingUser,
+    string $actingUsername,
     int $actingUserPermissions,
-    string $targetUser,
+    string $targetUsername,
     int $targetUserNewPermissions
 ): array {
     $retVal = [];
-    sanitize_sql_inputs($actingUser, $targetUser);
+    sanitize_sql_inputs($actingUsername, $targetUsername);
```
[the remark this answers is not in the document]
Comment (wescopeland) [reply app/Helpers/database/user-permission.php]: Thanks. In latest, I've moved the ORM query above sanitization so we can avoid this.

File: app/Helpers/database/user-permission.php [hunk 3]
```diff
@@ -71,27 +72,36 @@ function SetAccountPermissionsJSON(
...
+    // If the user is being unbanned, clear their `banned_at` timestamp.
+    if (
+        $targetUserCurrentPermissions < Permissions::Unregistered
+        && $targetUserNewPermissions >= Permissions::Unregistered
+    ) {
+        $userModel = User::where('User', $targetUsername)->first();
+        if ($userModel) {
+            $userModel->banned_at = null;
+            $userModel->save();
+        }
```
[the remark this answers is not in the document]
Comment (wescopeland) [reply app/Helpers/database/user-permission.php]: Good call, done in latest.

File: app/Helpers/database/user-permission.php [hunk 4]
```diff
@@ -19,48 +20,48 @@ function getUserPermissions(?string $user): int
...
+    string $targetUsername,
     int $targetUserNewPermissions
 ): array {
     $retVal = [];
-    sanitize_sql_inputs($actingUser, $targetUser);
+    sanitize_sql_inputs($actingUsername, $targetUsername);
 
-    $targetUserData = [];
-    if (!getAccountDetails($targetUser, $targetUserData)) {
+    $targetUser = User::where('User', $targetUsername)->first();
```
[the remark this answers is not in the document]
Comment (wescopeland) [reply app/Helpers/database/user-permission.php]: Updated in latest.

**Review by Jamiras — APPROVED** [review]

**Pull request merged**